controllers/oauth_client.go

```diff
@@ -0,0 +1,85 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"html/template"
+	"net/http"
+
+	service "go-crawler-challenge/services/oauth"
+
+	"github.com/beego/beego/v2/server/web"
+)
+
+// OauthClientController operations for oauth client generator
+type OauthClientController struct {
+	BaseController
+}
+
+// NestPrepare prepares some configurations to the controller
+func (c *OauthClientController) NestPrepare() {
+	c.actionPolicyMapping()
+}
+
+// URLMapping maps oauth client controller actions to functions
+func (c *OauthClientController) URLMapping() {
+	c.Mapping("New", c.New)
+	c.Mapping("Create", c.Create)
+}
+
+// actionPolicyMapping maps oauth client controller actions to policies
+func (c *OauthClientController) actionPolicyMapping() {
+	c.MappingPolicy("New", Policy{requireAuthenticatedUser: true})
+	c.MappingPolicy("Create", Policy{requireAuthenticatedUser: true})
+}
+
+// New handles a form for generating a new oauth client credential and only display into the same form
+// @Title New
+// @Description show an oauth client credential form with generated values
+// @Success 200
+// @router /oauth_client [get]
+func (c *OauthClientController) New() {
+	web.ReadFromRequest(&c.Controller)
+	flash := web.NewFlash()
+
+	clientId := c.GetString("client_id")
+	if clientId != "" {
+		oauthClient, err := service.ClientStore.GetByID(context.TODO(), clientId)
+		if err != nil {
+			flash.Error("OAuth client credential not found!")
+		} else {
+			c.Data["ClientId"] = oauthClient.GetID()
+			c.Data["ClientSecret"] = oauthClient.GetSecret()
+		}
+	}
+
+	flash.Store(&c.Controller)
+	c.Data["XSRFForm"] = template.HTML(c.XSRFFormHTML())
+	c.Layout = "layouts/application.html"
+	c.TplName = "oauth_client/new.html"
+}
+
+// Create handles create an oauth client credential
+// @Title Create
+// @Description create an oauth client credential
+// @Success 302 redirect to the oauth client form path with success message
+// @Success 302 redirect to the oauth client form path with error message
+// @router /oauth_client [post]
+func (c *OauthClientController) Create() {
+	flash := web.NewFlash()
+	redirectPath := "/oauth_client"
+
+	domain := c.Ctx.Request.Host
```

Comment:
The domain actually should be client's domain, not the oauth server domain. But as we will be using the grant type password for authentication(not the others grant type where it needed callbacks url), I think we dont need to even set this client domain.

Reply:
Removed